### PR TITLE
[AC-5345] Winner Statuses should be available in Salesforce

### PR DESCRIPTION
**Pull request by knittingarch**
**STEP 2: This step creates the history event.  The creation of the Startup Roles and the assignment of them to the winners happens in STEP 1 on accelerate.**

**To Test On The Branch**
- Run `make build; make dev`.
- In a separate tab, run `make load-db GZ_FILE=../accelerate/db_cache/initial_schema.sql.gz`.
- Run `make grant-permissions PERMISSION_USER=demoadmin@masschallenge.org PERMISSION_CLASSES=v1_clients`.
- In the browser, navigate to [http://localhost:8000/api/v1/organization/1707/history/](http://localhost:8000/api/v1/organization/1707/history/).
- Check that the `became winner` event shows for this user and that the datetime is the same as the end_date for the program (the startup status doesn't have a credible created_at date).
- You can use the list of startups from accelerate to test other organization's `became winner` event.

### CODE REVIEW

File: web/impact/impact/v1/events/organization_became_winner_event.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+                     "icontains"],
+                    "winner")):
+                result.append(cls(ss))
+        return result
+
+    def calc_datetimes(self):
+        self.earliest = self.startup_status.created_at
+        program_end_date = self._program.end_date
+        if self.earliest is None:
+            self.earliest = program_end_date
```
Comment (yotammanor) [inline web/impact/impact/v1/events/organization_became_winner_event.py]: Can be re-written as a one-liner:
`self.earliest = self.startup_status.created_at or self._program.end_date`

File: web/impact/impact/v1/events/organization_became_winner_event.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        return self._program.id
+
+    def winner_level(self):
+        return self.startup_status.program_startup_status.startup_role.name
+
+    def winner_level_name(self):
+        pss = self.startup_status.program_startup_status
+        program = pss.program.name
+        winner_level = pss.startup_role.name
+        return program + " " + winner_level
```
Comment (yotammanor) [inline web/impact/impact/v1/events/organization_became_winner_event.py]: I'd prefer a modern string formatting, either `'{} {}'.format(program, winner_level)`, or better yet, a marvellous f-string: `f'{program} {winner_level}'` !

**Review by yotammanor — APPROVED** [review]
Great PR. Works as expected, code looks great.

Two small comments, non-blocking as far as I'm concerned.

File: web/impact/impact/v1/events/organization_became_winner_event.py (hunk 2 above)
Comment (jonkiparsky) [reply web/impact/impact/v1/events/organization_became_winner_event.py]: Are f-strings marvelous? I mean, they seem sort of grotty to me - sort of antipythonic, if you like to put it that way. 
But yeah, string formatting > concatenation.

File: web/impact/impact/models/startup_role.py [hunk 3]
```diff
@@ -15,11 +15,26 @@ class StartupRole(MCModel):
     FAST_TRACK = "Fast Track"
     AIR = "Alum In Residence"
     MC_STAFF = "MC Staff"
+    WINNER = "Winner"
+    GOLD_WINNER = "Gold Winner"
+    SILVER_WINNER = "Silver Winner"
+    PLATINUM_WINNER = "Platinum Winner"
+    DIAMOND_WINNER = "Diamond Winner"
+    INKIND_WINNER = "In-Kind Winner"
```
Comment (jonkiparsky) [inline web/impact/impact/models/startup_role.py]: Should match the constant defined in accelerate :)

File: web/impact/impact/tests/test_organization_history_view.py [hunk 4]
```diff
@@ -333,6 +334,45 @@ def test_startup_became_finalist_no_created_at(self):
...
             self.assertEqual(start_date, events[0]["datetime"])
 
+    def test_startup_became_winner(self):
+        startup = StartupFactory()
+        startup_status = StartupStatusFactory(
+            program_startup_status__startup_role__name=StartupRole.SILVER_WINNER,
+            startup=startup)
+        program = startup_status.program_startup_status.program
+        role = startup_status.program_startup_status.startup_role
+        with self.login(email=self.basic_user().email):
```
Comment (jonkiparsky) [inline web/impact/impact/tests/test_organization_history_view.py]: why is this in a `with`? It doesn't seem like you need this to be a context block, as there's no cleanup to do and no exceptions to catch.

File: web/impact/impact/tests/test_organization_history_view.py [hunk 5]
```diff
@@ -333,6 +334,45 @@ def test_startup_became_finalist_no_created_at(self):
...
+            url = reverse(OrganizationHistoryView.view_name,
+                          args=[startup.organization.id])
+            response = self.client.get(url)
+            events = find_events(response.data["results"],
+                                 OrganizationBecameWinnerEvent.EVENT_TYPE)
+            self.assertEqual(1, len(events))
```
Comment (jonkiparsky) [inline web/impact/impact/tests/test_organization_history_view.py:349]: This test seems to be making about half a dozen assertions. We generally prefer to test one thing at a time. What exactly are we trying to test here?

File: web/impact/impact/tests/test_organization_history_view.py [hunk 6]
```diff
@@ -333,6 +334,45 @@ def test_startup_became_finalist_no_created_at(self):
...
+    def test_startup_became_winner_no_created_at(self):
+        startup = StartupFactory()
+        end_date = days_from_now(-10)
+        startup_status = StartupStatusFactory(
+            program_startup_status__startup_role__name=StartupRole.GOLD_WINNER,
+            program_startup_status__program__end_date=end_date,
+            startup=startup)
+        startup_status.created_at = None
+        startup_status.save()
+        with self.login(email=self.basic_user().email):
```
Comment (jonkiparsky) [inline web/impact/impact/tests/test_organization_history_view.py:361]: likewise, why a `with`?

**Review by jonkiparsky — CHANGES_REQUESTED** [review]

File: web/impact/impact/v1/events/organization_became_winner_event.py (hunk 2 above)
Comment (knittingarch) [reply web/impact/impact/v1/events/organization_became_winner_event.py]: Fair enough.  LOL!  I'll switch this to regular string formatting :)

File: web/impact/impact/models/startup_role.py (hunk 3 above)
Comment (knittingarch) [reply web/impact/impact/models/startup_role.py]: Hahahah!  Roger that!

File: web/impact/impact/tests/test_organization_history_view.py (hunk 6 above)
Comment (knittingarch) [reply web/impact/impact/tests/test_organization_history_view.py:361]: I was trying to keep the test consistent with the ones written previously.

File: web/impact/impact/models/startup_role.py (hunk 3 above)
Comment (knittingarch) [reply web/impact/impact/models/startup_role.py]: Done!

File: web/impact/impact/tests/test_organization_history_view.py (hunk 6 above)
Comment (jonkiparsky) [reply web/impact/impact/tests/test_organization_history_view.py:361]: Oh, okay, I see. It's that TestPlus nonsense. Okay, not your fault.

File: web/impact/impact/tests/test_organization_history_view.py (hunk 4 above)
Comment (knittingarch) [reply web/impact/impact/tests/test_organization_history_view.py]: As with my other comment, I wanted this test to be consistent with the previously written tests.  The only line that I am really interested in is `self.assertEqual(role.name, events[0]["winner_level"])`.  Can strip the rest of the things out if you prefer.

File: web/impact/impact/v1/events/organization_became_winner_event.py (hunk 1 above)
Comment (knittingarch) [reply web/impact/impact/v1/events/organization_became_winner_event.py]: Done!

File: web/impact/impact/tests/test_organization_history_view.py (hunk 5 above)
Comment (knittingarch) [reply web/impact/impact/tests/test_organization_history_view.py:349]: See previous comment.

**Review by jonkiparsky — APPROVED** [review]

**Pull request merged**